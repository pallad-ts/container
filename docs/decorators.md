# Decorators
Creating service definition via container (`definitionWithConstructor` etc) might be a tedious and error-prone work.
That's why _alpha-dic_ supports decorators for classes.

NOTE! Hopefully this is obvious but Decorators are not supported by node.js natively. You need to use Typescript (highly recommended) or other transpiler (like Babel) to make it work.

```typescript
const {Service, Annotation, Inject} = require('alpha-dic');

@Service() // you can also provide a name @Service('name'). By default class/function name inferred.
@Annotation({name: 'extraAnnotation'})
export class NotificationCenter {
    
    @Inject('superExtraService')
    public extraService: ExtraService;
    
    constructor(
        @Inject('EmailRenderer') emailRenderer: EmailRenderer, 
        @Inject('EmailSender') emailSender: EmailSender) {
    }
}

```

## Setup

Since decorator is just regular function you need to give it an opportunity to get called. To do soo you can use `loadServices` function that loads all modules matching given patterns.  
```typescript
import {createStandard, loadServices} from 'alpha-dic';

const container = createStandard();

// loads all files (with extensions .ts, .js, .tsx but ignores .d.ts, .d.tsx) in given directory so all @Service decorator calls got a chance to register service definitions
loadServices(container, {
    patterns: './services/*.*' // note ".*" to make sure glob can catch all files
}); 
```