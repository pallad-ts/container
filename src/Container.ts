import {
    AnnotationPredicate, ServiceName, DefinitionPredicate, Middleware,
    ServiceFactory
} from './types';
import {Definition} from './Definition';
import * as errors from './errors';
import {assertNoCircularDependencies} from './assertNoCircularDependencies';
import {ContainerArg} from './ContainerArg';
import * as is from 'predicates';
import {randomName} from "./randomName";

function isThenable(result: any): result is Promise<any> {
    return result && 'then' in result;
}

export class Container {
    private definitions: Map<string | Symbol, Definition> = new Map();
    private services: Map<Definition, Promise<any>> = new Map();
    private middlewares: Middleware[] = [];

    public readonly parent: Container;

    constructor(parent?: Container) {
        Object.defineProperty(this, 'parent', {
            value: parent,
            writable: false
        });
    }

    /**
     * Registers given service definition
     */
    registerDefinition(definition: Definition): this {
        if (this.definitions.has(definition.name)) {
            throw errors.ALREADY_DEFINED(`Service "${definition.name.toString()}" already defined`);
        }
        this.definitions.set(definition.name, definition);
        return this;
    }

    /**
     * Creates and registers service definition
     *
     * Returns created definition for further configuration
     */
    definition(name?: ServiceName) {
        const definition = new Definition(name);
        this.registerDefinition(definition);
        return definition;
    }

    /**
     * Creates and registers service definition with given name, function as constructor
     */
    definitionWithConstructor(name: ServiceName, clazz: Function): Definition;
    definitionWithConstructor(clazz: Function): Definition;
    definitionWithConstructor(nameOrClazz: ServiceName | Function, clazz?: Function) {
        const finalClazz: Function = is.func(nameOrClazz) ? nameOrClazz : clazz!;
        return this.definition(ServiceName.is(nameOrClazz) ? nameOrClazz : randomName(finalClazz.name))
            .useConstructor(finalClazz);
    }

    /**
     * Creates and registers service definition with given name, function as factory
     */
    definitionWithFactory(name: ServiceName, factory: ServiceFactory, type?: Function): Definition;
    definitionWithFactory(factory: ServiceFactory, type?: Function): Definition;
    definitionWithFactory(nameOrFactory: ServiceName | ServiceFactory, factoryOrType: ServiceFactory | Function, type?: Function): Definition {
        const name = ServiceName.is(nameOrFactory) ? nameOrFactory : undefined;
        const factory = is.func(nameOrFactory) ? nameOrFactory as ServiceFactory : factoryOrType as ServiceFactory;
        const finalType = is.func(nameOrFactory) ? factoryOrType : type;

        const def = this.definition(name)
            .useFactory(factory);
        if (finalType) {
            def.markType(finalType);
        }
        return def;
    }

    /**
     * Creates and registers service definition with given name and value as a service
     */
    definitionWithValue(name: ServiceName, value: any): Definition;
    definitionWithValue(value: any): Definition;
    definitionWithValue(nameOrValue: ServiceName | any, value?: any): Definition {
        const isNameProvided = value !== undefined && ServiceName.is(nameOrValue);
        const finalValue = isNameProvided ? value : nameOrValue;
        const name = isNameProvided ? nameOrValue : randomName(Object.getPrototypeOf(finalValue).constructor.name);
        return this.definition(name)
            .useValue(finalValue);
    }

    /**
     * Returns definition by given name
     */
    findByName(name: ServiceName): Definition | undefined {
        if (this.definitions.has(name)) {
            return this.definitions.get(name);
        }

        if (this.parent) {
            return this.parent.findByName(name);
        }
    }

    /**
     * Returns definitions that satisfy given predicate
     */
    findByPredicate(predicate: DefinitionPredicate): Definition[] {
        return Array.from(this.definitions.values())
            .filter(predicate)
            .concat(this.parent ? this.parent.findByPredicate(predicate) : []);
    }


    /**
     * Returns all definitions that contain annotation that satisfied given predicate
     */
    findByAnnotation(predicate: AnnotationPredicate): Definition[];
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: false): Definition[];
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: true): [Definition, any][];
    findByAnnotation(predicate: AnnotationPredicate, withAnnotation: boolean = false): Definition[] | [Definition, any][] {
        let annotations: any[] = [];
        const definitions = this.findByPredicate(s => {
            const annotation = s.annotations.find(predicate);

            if (annotation) {
                withAnnotation && annotations.push(annotation);
                return true;
            }
            return false;
        });

        if (withAnnotation) {
            return definitions.map((d): [Definition, any] => {
                return [d, annotations.shift()];
            });
        }
        return definitions;
    }

    /**
     * Registers given middleware
     */
    addMiddleware(...middlewares: Middleware[]): Container {
        this.middlewares.push(...middlewares);
        return this;
    }

    getMiddlewares(): Middleware[] {
        return this.middlewares;
    }

    /**
     * Returns service for given name or definition
     */
    get<T = any>(nameOrDefinition: ServiceName | Definition): Promise<T> {
        let definition: Definition | undefined;

        if (ServiceName.is(nameOrDefinition)) {
            definition = this.findByName(nameOrDefinition);
            if (definition === undefined) {
                return Promise.reject(errors.SERVICE_NOT_FOUND(`Service "${nameOrDefinition.toString()}" does not exist`));
            }
        } else {
            definition = nameOrDefinition;
        }

        if (this.services.has(definition)) {
            return this.services.get(definition)!;
        }

        if (!this.definitions.has(definition.name) && this.parent) {
            return this.parent.get(definition);
        }

        const promise = this.create(definition);
        this.services.set(definition, promise);
        return promise;
    }

    private create(definition: Definition) {
        if (!definition.factory) {
            return Promise.reject(
                errors.INCOMPLETE_DEFINITION(
                    `Missing factory for service definition "${definition.name.toString()}". Define it as constructor, factory or value`
                )
            );
        }

        try {
            assertNoCircularDependencies(this, definition);
        } catch (e) {
            return Promise.reject(e);
        }

        // valid definition, time to lock it
        definition.lock();

        let currentMiddleware = 0;
        const middlewares = this.getMiddlewares();
        const next = (definition: Definition) => {
            const middleware = middlewares[currentMiddleware];
            currentMiddleware++;
            if (middleware) {
                return middleware.call(this, definition, next);
            } else {
                return Promise.all(
                    definition.args.map(a => a instanceof ContainerArg ? a.getArgument(this) : a)
                )
                    .then((args: any[]) => definition.factory.apply(this, args));
            }
        };
        const result = next.call(this, definition);
        return isThenable(result) ? result : Promise.resolve(result);
    }

    /**
     * Returns all services that definition satisfies predicate
     */
    getByPredicate<T = any>(predicate: DefinitionPredicate): Promise<T[]> {
        return Promise.all(
            this.findByPredicate(predicate)
                .map(d => this.get(d))
        );
    }

    /**
     * Returns all services that definition contains annotation that satisfies given predicate
     */
    getByAnnotation<T = any>(predicate: AnnotationPredicate): Promise<T[]>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: false): Promise<T[]>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: true): Promise<[T, any][]>;
    getByAnnotation<T = any>(predicate: AnnotationPredicate, withAnnotation: boolean = false): Promise<T[] | [T, any][]> {
        return Promise.all(
            this.findByAnnotation(predicate, true)
                .map(async ([definition, annotation]) => {
                    if (withAnnotation) {
                        return [await this.get(definition), annotation];
                    }
                    return this.get(definition);
                })
        );
    }
}