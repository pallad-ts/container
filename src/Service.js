'use strict';
/**
 * @namespace AlphaDIC
 */

/**
 * @class
 */
class Service {
    constructor(name) {
        /**
         * @member {string}
         */
        this.name = name;
        /**
         * @type {boolean}
         */
        this.cacheable = true;
        /**
         * @type {Object}
         */
        this.annotations = Object.create(null);
        /**
         * @type {Array<string>}
         */
        this.dependencies = [];
    }

    /**
     * Makes service non cacheable which means it will be created every time
     */
    nonCacheable() {
        this.cacheable = false;
    }

    /**
     * Sets constructor value used to create instance of service.
     *
     * @param {Function} constructorFunction
     * @returns {Service}
     */
    useConstructor(constructorFunction) {
        this.type = Service.TYPE_CONSTRUCTOR;
        this.value = constructorFunction;
        return this;
    }

    /**
     * Sets factory value used to create instance of service.
     * The value should return promise in case of asynchronous service creation.
     * The factory value is called in context of AlphaDIC.
     *
     * @param {Function} factoryFunction
     * @returns {Service}
     */
    useFactory(factoryFunction) {
        this.type = Service.TYPE_FACTORY;
        this.value = factoryFunction;
        return this;
    }

    /**
     * Sets given argument as a value for the service.
     * The service will be always resolved to given value.
     *
     *
     * @param {*} value
     * @returns {Service}
     */
    useValue(value) {
        this.type = Service.TYPE_VALUE;
        this.value = value;
        return this;
    }

    /**
     * Sets asynchronous factory value used to create instance of service.
     * The value should call callback provided as last argument to the value (after dependencies)
     * The factory value is called in context of AlphaDIC.
     *
     * @example
     * const service = new Service('service')
     * service.useAsyncFactory((A, callback) => {
     *      setTimeout(() => {
     *          callback(null, {A: A, some: 'property'});
     *      }, 100);
     * });
     *
     * @param {Function} factoryFunction
     * @returns {Service}
     */
    useAsyncFactory(factoryFunction) {
        this.type = Service.TYPE_ASYNC_FACTORY;
        this.value = factoryFunction;
        return this;
    }

    /**
     * Defines dependencies that will be resolved and then injected to the service.
     *
     * @param {(string|Array<string>|...string)} services
     */
    dependsOn(services) {
        if (Array.isArray(services)) {
            this.dependencies = services;
        } else {
            this.dependencies = Array.from(arguments);
        }
        return this;
    }

    /**
     * @typedef {Object} Annotation
     * @property {string} name
     */

    /**
     * Sets annotation with given name and properties
     *
     * @param {(Annotation|string)} name
     * @param {Object} [properties={}]
     * @returns {Service}
     */
    annotate(name, properties) {
        if (!name) {
            throw new Error('Annotation name has to be string or annotation object');
        }

        if (typeof name === 'string') {
            this.annotations[name] = properties || {};
        } else {
            if (!name.name) {
                throw new Error('Annotation object requires non-empty "name" property');
            }
            this.annotations[name.name] = name;
        }
        return this;
    }

    /**
     * Checks whether given annotation exists
     *
     * @param {string} name
     * @returns {boolean}
     */
    hasAnnotation(name) {
        return name in this.annotations;
    }

    /**
     * Returns properties for given annotation
     *
     * @param {string} name
     * @returns {Object}
     */
    getAnnotation(name) {
        return this.annotations[name];
    }

    /**
     * Returns all annotations
     *
     * @returns {Object}
     */
    getAnnotations() {
        return this.annotations;
    }
}

Service.TYPE_CONSTRUCTOR = 'constructor';
Service.TYPE_FACTORY = 'factory';
Service.TYPE_ASYNC_FACTORY = 'async-factory';
Service.TYPE_VALUE = 'value';

module.exports = Service;