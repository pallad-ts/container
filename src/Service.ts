export interface Annotation {
    name: string;
    [extraProperty: string]: any
}

enum ServiceCreationType {
    CONSTRUCTOR,
    FACTORY,
    VALUE
}

export class Service {
    public cacheable = true;
    public annotations = new Map<string, Annotation>();
    public dependencies: string[] = [];

    private creationType: ServiceCreationType;
    private creator: () => Promise<any>;

    constructor(public readonly name: string) {

    }

    /**
     * Makes service non cacheable which means it will be created every time
     *
     * @returns {Service}
     */
    nonCacheable(): this {
        this.cacheable = false;
        return this;
    }

    /**
     * Sets constructor of service to use when creating an instance of it
     */
    useConstructor(func: Function): this {
        this.creationType = ServiceCreationType.CONSTRUCTOR;
        this.creator = func;
        return this;
    }

    /**
     * Alias for {@see useConstructor}
     */
    useClass(clazz: Function) {
        return this.useConstructor(clazz);
    }

    /**
     * Sets factory used to create an instance of service.
     * In case of asynchronours service creation, factory should return a promise.
     * The factory value is called in context of AlphaDIC.
     */
    useFactory(factory: () => any | Promise<any>) {
        this.creationType = ServiceCreationType.FACTORY;
        this.creator = factory;
        return this;
    }

    useValue(value) {
        this.creationType = ServiceCreationType.VALUE;
        this.creator = value;
        return this;
    }

    /**
     * Adds annotation to the service
     * Annotation is a simple metadata object assigned to service that you might use for different purposes
     */
    annotate(name: string): this;
    annotate(name: string, properties: object): this;
    annotate(annotation: Annotation): this;
    annotate(nameOrAnnotation: string | Annotation, properties?: object): this {
        if (!nameOrAnnotation) {
            throw new Error('Annotation name has to be string or annotation object');
        }

        let annotation: Annotation;
        if (typeof nameOrAnnotation === 'string') {
            annotation = Object.assign({}, properties, {name: nameOrAnnotation});
        } else {
            if (!nameOrAnnotation.name) {
                throw new Error('Annotation object requires non-empty "name" property');
            }
            annotation = nameOrAnnotation;
        }
        this.annotations.set(annotation.name, annotation);
        return this;
    }

    /**
     * Dependencies to inject to the service
     */
    dependsOn(services: string[]);
    dependsOn(...services: string[]);
    dependsOn(services: string[] | string, ...extraServices: string[]) {
        if (Array.isArray(services)) {
            this.dependencies = services;
        } else {
            this.dependencies = (<string[]>services).concat(extraServices);
        }
        return this;
    }
}