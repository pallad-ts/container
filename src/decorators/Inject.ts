import {ContainerArg} from "../ContainerArg";
import 'reflect-metadata';
import {Reference} from "../Reference";
import * as is from 'predicates';
import {ensureMetadata} from "../serviceMetadata";
import {ServiceName} from "../types";
import {TypeRef} from "../TypeRef";
import {errors} from "../index";
import {format} from 'util';

const assertServiceNameOrContainerArg = is.assert(
    is.any(ServiceName.is, ContainerArg.is, TypeRef.is, is.func, is.undefined),
    '@Inject argument must be a string that represents service name, symbol, function or instance of ContainerArg or TypeRef'
);

export function Inject(ref: ServiceName | ContainerArg | Function): ParameterDecorator & PropertyDecorator;
export function Inject(): PropertyDecorator;
export function Inject(ref?: ServiceName | ContainerArg | Function): PropertyDecorator | ParameterDecorator {
    assertServiceNameOrContainerArg(ref);

    return function (target: any, property: string | symbol, indexOrDescriptor?: number | TypedPropertyDescriptor<any>) {
        const isParameterDecorator = typeof indexOrDescriptor === 'number';

        let arg: Reference | ContainerArg;

        if (ref === undefined) {
            if (isParameterDecorator) {
                throw new Error('Using @Inject decorator without ref for function arguments is prohibited');
            } else {
                const designType = Reflect.getMetadata("design:type", target, property);
                arg = createReferenceForType(designType, `property ${property.toString()}`);
            }
        } else if (TypeRef.is(ref)) {
            arg = Reference.one.type(ref);
        } else if (ServiceName.is(ref)) {
            arg = Reference.one.name(ref);
        } else if (ContainerArg.is(ref)) {
            arg = ref;
        } else {
            arg = createReferenceForType(
                ref,
                isParameterDecorator ? `constructor (of ${target.name}) argument nr: ${indexOrDescriptor}` : `property ${property.toString()}`
            );
        }

        if (isParameterDecorator) {
            ensureMetadata(target).constructorArguments[<number>indexOrDescriptor] = arg!;
        } else {
            ensureMetadata(target.constructor).propertiesInjectors.set(property, arg!);
        }
    }
}

function createReferenceForType(func: Function, ...args: string[]) {
    const type = TypeRef.createFromType(func);
    if (type === undefined) {
        throw errors.AUTOWIRING_FAILED(
            format(errors.AUTOWIRING_FAILED.defaultMessage, ...args)
        );
    }
    return Reference.one.type(type);
}