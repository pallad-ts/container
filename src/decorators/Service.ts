import {Container} from "../Container";
import {ensureMetadata, createDefinitionFromMetadata} from "../serviceMetadata";

import 'reflect-metadata';
import {Definition} from "../Definition";
import {ServiceName} from "../types";
import {randomName} from "../randomName";

export interface ServiceType {
    (name?: ServiceName): ClassDecorator;

    useContainer(container?: Container): void;

    _container?: Container
}

const DEFINITION_KEY = Symbol('__alphaDic-ServiceDefinition');
export const Service = function (name?: ServiceName) {
    return function (constructor: {new(...args: any[]): any}) {
        const finalName = name || randomName(constructor.name);

        const metadata = ensureMetadata(constructor);
        metadata.name = finalName;

        const definition = createDefinitionFromMetadata(metadata, constructor);
        Reflect.defineMetadata(DEFINITION_KEY, definition, constructor);

        if (Service._container) {
            Service._container.registerDefinition(definition);
        } else if (!process.env.ALPHA_DIC_NO_SERVICE_CONTAINER) {
            console.warn('There is no container registered in @Service decorator. ' +
                'Use Service.useContainer(container) to define the container. ' +
                'Without it service definitions created via @Service decorator cannot be registered automatically. ' +
                'If you wish to register service definitions manually via getDefinitionForClass set ALPHA_DIC_NO_SERVICE_CONTAINER environment variable to disable this warning.')
        }
    }
} as ServiceType;

Service.useContainer = function (container: Container) {
    Service._container = container;
};

export function getDefinitionForClass(clazz: Function): Definition {
    return Reflect.getMetadata(DEFINITION_KEY, clazz);
}
