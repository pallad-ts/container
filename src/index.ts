import * as errors from './errors';
import {Container} from './Container';

import {
    onActivation as onActivationAnnotation,
    activationMiddleware as _activationMiddleware
} from './middlewares/activation';

export * from './Definition';
export * from './Container';
export * from './Reference';
export * from './types';
export * from './ContainerArg';
export * from './referenceFunc';
export * from './Lookup';

export {errors};

export function create() {
    return new Container();
}

export const activationMiddleware = _activationMiddleware;
export const onActivation = onActivationAnnotation;
export {Service} from './decorators/Service';
export {Annotation} from './decorators/Annotation';
export {Inject} from './decorators/Inject';