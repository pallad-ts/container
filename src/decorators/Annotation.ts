import {ensureMetadata} from "./serviceMetadata";
import {getDefinitionForClass} from "./Service";

export function Annotation(annotation: any): ClassDecorator {
    return (clazz: Function) => {
        const definition = getDefinitionForClass(clazz);
        if (definition) {
            definition.annotate(annotation);
        }
        ensureMetadata(clazz).annotations.push(annotation);
    }
}