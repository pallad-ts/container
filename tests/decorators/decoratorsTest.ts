import {getDefinitionForClass, Service} from "../../src/decorators/Service";
import {assert} from 'chai';
import {Inject} from "../../src/decorators/Inject";
import {Reference} from "../../src/Reference";
import {Container} from "../../src/Container";
import {Annotation} from "../../src/decorators/Annotation";
import {SinonStub} from "sinon";
import * as sinon from 'sinon';

describe('decorators', () => {
    it('simple service', () => {
        @Service('SomeService')
        class Test {

        }

        const definition = getDefinitionForClass(Test);
        assert.propertyVal(definition, 'name', 'SomeService');

        const result = definition.factory(...definition.args);
        assert.instanceOf(result, Test);
    });

    it('simple service without explicit name', () => {
        @Service()
        class TestFoo {

        }

        const definition = getDefinitionForClass(TestFoo);
        expect(definition.name)
            .toMatch(/^TestFoo.*/);
    });

    it('service with injected args', () => {
        @Service()
        class Foo {
            constructor(@Inject('a') arg1: any) {

            }
        }

        const definition = getDefinitionForClass(Foo);
        expect(definition.name)
            .toMatch(/^Foo.*/);

        assert.sameDeepMembers(definition.args, [
            Reference.one.name('a')
        ]);
    });

    it('Injecting advanced reference', () => {
        const ref = Reference.one.annotation(() => true);

        @Service()
        class Foo {
            constructor(@Inject(ref) arg1: any) {

            }
        }

        const definition = getDefinitionForClass(Foo);
        assert.sameMembers(definition.args, [ref]);
    });

    it('Inject accepts only string or object of ContainerArg class', () => {
        expect(() => {
            @Service()
            class Foo {
                constructor(@Inject([] as any) arg1: any) {

                }
            }
        })
            .toThrowErrorMatchingSnapshot();
    });

    it('service with injected args and properties', () => {
        @Service()
        class Foo {
            @Inject('bar')
            bar: any;

            constructor(@Inject('a') private arg1: any, @Inject('b') private arg2: any) {

            }
        }

        const definition = getDefinitionForClass(Foo);
        expect(definition.name)
            .toMatch(/^Foo.*/);

        assert.sameDeepMembers(definition.args, [
            Reference.one.name('a'),
            Reference.one.name('b'),
            Reference.one.name('bar')
        ]);

        const result = definition.factory('a', 'b', 'bar');

        assert.instanceOf(result, Foo);
        assert.propertyVal(result, 'bar', 'bar');
        assert.propertyVal(result, 'arg1', 'a');
        assert.propertyVal(result, 'arg2', 'b');
    });

    it('fails if not all constructor arguments have @Inject decorators', () => {
        assert.throws(() => {
            @Service()
            class Foo {
                constructor(@Inject('a') arg1: any, arg2: any) {

                }
            }
        }, /Required constructor arguments: 2, provided: 1/);
    });

    it('fails it there is an argument, in the middle of other arguments, without @Inject decorator', () => {
        assert.throws(() => {
            @Service()
            class Foo {
                constructor(@Inject('a') arg1: any, arg2: any, @Inject('c') arg3: any) {

                }
            }
        }, /Missing @Inject decorator for argument at position "1"/)
    });

    describe('using container', () => {
        let container: Container;
        beforeEach(() => {
            container = new Container();
            Service.useContainer(container);
        });

        afterEach(() => {
            Service.useContainer(undefined);
        });

        it('set container via Service.useContainer()', () => {

            @Service('Test')
            class Test {

            }

            assert.strictEqual(container.findByName('Test'), getDefinitionForClass(Test));
        });
    });

    it('Defining annotation', () => {
        const annotation1 = {name: 'test'}
        const annotation2 = {name: 'test2'};

        @Annotation(annotation1)
        @Annotation(annotation2)
        @Service()
        class Test {

        }

        const definition = getDefinitionForClass(Test);
        assert.sameMembers(definition.annotations, [annotation1, annotation2]);
    });

    it('defining annotation when @Service decorator is used first', () => {
        const annotation1 = {name: 'test'}
        const annotation2 = {name: 'test2'};

        @Service()
        @Annotation(annotation1)
        @Annotation(annotation2)
        class Test {

        }

        const definition = getDefinitionForClass(Test);
        assert.sameMembers(definition.annotations, [annotation1, annotation2]);
    });

    describe('Service warning is no container registered', () => {
        let consoleWarnStub: SinonStub;
        beforeEach(() => {
            delete process.env.ALPHA_DIC_NO_SERVICE_CONTAINER;
            consoleWarnStub = sinon.stub(console, 'warn');
        });

        afterEach(() => {
            process.env.ALPHA_DIC_NO_SERVICE_CONTAINER = '1';
            consoleWarnStub.restore();
        });

        it('test', () => {
            @Service()
            class Foo {

            }

            sinon.assert.calledWithMatch(consoleWarnStub, sinon.match(/There is no container registered in @Service decorator/));
        });
    })
});