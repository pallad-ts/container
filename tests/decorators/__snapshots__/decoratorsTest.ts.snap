// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`decorators Inject accepts only string or object of ContainerArg class 1`] = `"@Inject argument must be a string that represents service name, symbol, function or instance of ContainerArg or TypeRef"`;
